qiskit_experiments/calibration/update_library.py

```diff
@@ -0,0 +1,174 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A library of experiment calibrations."""
+
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import List, Tuple, Union
+import numpy as np
+
+from qiskit.circuit import Parameter
+from qiskit.pulse import ScheduleBlock
+
+from qiskit_experiments.experiment_data import ExperimentData
+from qiskit_experiments.calibration.backend_calibrations import BackendCalibrations
+from qiskit_experiments.calibration.calibrations import Calibrations
+from qiskit_experiments.calibration.parameter_value import ParameterValue
+from qiskit_experiments.calibration.exceptions import CalibrationError
+from qiskit_experiments.calibration.calibration_key_types import ParameterValueType
+
+
+class BaseUpdater(ABC):
+    """A base class to update calibrations."""
+
+    @staticmethod
+    def _time_stamp(exp_data: ExperimentData) -> datetime:
+        """Helper method to extract the datetime."""
+        all_times = exp_data.completion_times.values()
+        if all_times:
+            return max(all_times)
+
+        return datetime.now()
+
+    @classmethod
+    def _update(
+        cls,
+        exp_data: ExperimentData,
+        cal: Calibrations,
+        value: ParameterValueType,
+        param: Union[Parameter, str],
+        schedule: Union[ScheduleBlock, str] = None,
+        group: str = "default",
+    ):
+        """Update the calibrations with the given value.
+
+        Args:
+            exp_data: The ExperimentData instance that contains the result and the experiment data.
+            cal: The Calibrations instance to update.
+            value: The value extracted by the subclasses in the :meth:`update` method.
+            param: The name of the parameter, or the parameter instance, which will receive an
+                updated value.
+            schedule: The ScheduleBlock instance or the name of the instance to which the parameter
+                is attached.
+            group: The calibrations group to update.
+        """
+
+        qubits = exp_data.data(0)["metadata"]["qubits"]
+
+        param_value = ParameterValue(
+            value=value,
+            date_time=BaseUpdater._time_stamp(exp_data),
+            group=group,
+            exp_id=exp_data.experiment_id,
+        )
+
+        cal.add_parameter_value(param_value, param, qubits, schedule)
+
+    @classmethod
+    @abstractmethod
+    def update(cls, exp_data: ExperimentData, calibrations: BackendCalibrations, **options):
```

Comment:
Is it possible to swap out `_update` and `update`? If `_update` is the method that defines the behavior of cal value extraction, we can avoid explicitly calling current `_update` in every subclasses (i.e. if a developer doesn't know `_update` method they might write this by themself, which is not expected situation). 

Something like:

```python
def update(calibrations, data, **some_kwargs):
    # do prep
    cal_params = self._update(data, **some_kwargs)
    for cal_param in cal_params:
        # add cal_param to calibrations
```

Reply:
I don't think that makes sense `_update` is the method that updates the parameter value but does not define how the calibrations are extracted. It is however confusing to have both `_update` and `update`. I therefore kept `update` but renamed `_update` to `_add_parameter_value` which better reflects its intent. I also improved the docstring for `update` in the base class as this is what the developer will have to implement. See 8e65eef.